### PR TITLE
Removed forName method 

### PR DESCRIPTION
**Pull request by martin-morek**
- fixes [686](https://github.com/ably/ably-java/issues/686) and [685](https://github.com/ably/ably-java/issues/685)
- Removed use of `forName` method in push activation state machine implementation and replaced by switch constructing Event object based on name

### CODE REVIEW

File: android/src/main/java/io/ably/lib/push/ActivationStateMachine.java [hunk 1]
```diff
@@ -678,21 +717,12 @@ private boolean persist() {
...
-                    continue;
-                }
+                ActivationStateMachine.Event event = Event.constructEventByName(className);
                 deque.add(event);
-            } catch(Exception e) {
-                throw new RuntimeException(e);
+            } catch (ClassNotFoundException e) {
+                Log.e(TAG, e.getLocalizedMessage());
+            } catch (InstantiationException e) {
+                continue;
```
Comment (QuintinWillison) [inline android/src/main/java/io/ably/lib/push/ActivationStateMachine.java:725]: If this needs to be checked then is there any reason we're not logging it out also?

**Review by QuintinWillison — CHANGES_REQUESTED** [review]
Thanks, this is a great start. However, as discussed in our call earlier today, this only presents one side of the story... I think we need to see:

- Addition of functionality that does directly the opposite of your `constructEventByName` static method - probably an instance method on `Event`, perhaps simply something like `getEventName()`
- Some unit tests for this class hierarchy proving:
    - that `Event` instances can be represented as string and then rehydrated from the same
    - that attempts to rehydrate unsupported event types (those with non-nullary constructor) throw the right exception
    - that attempts to rehydrate unknown event types throw the right exception

We need to get rid of any code that uses runtime reflection. This has to be checked by the compiler.

**Comment by martin-morek** [comment]
@QuintinWillison could you please elaborate on `getEventName()` method? Right now for getting event name we are using `getClass().getName()`, I'm not sure if this can be simplified. What we can do is define property called `name` for an event object and manually fill it with a value. That way when we change class name code stays functional but real name would differ.

**Comment by QuintinWillison** [comment]
Hi @martin-morek - so, the use of `getClass().getName()` is the crux of the issue! 😄 

That method is dynamic at runtime. If the class name is different, because of obfuscation or something else outside of control as SDK developers, then that can mess things up - in particular from build to build of a application.

Each class mast vend its own serialisation name from a static string. These cannot be dynamically sourced like this.

File: android/src/main/java/io/ably/lib/push/ActivationStateMachine.java [hunk 2]
```diff
@@ -26,169 +26,200 @@
...
-        public abstract String getName();
-
-        public static Event constructEventByName(String className) throws ClassNotFoundException, InstantiationException {
-            ActivationStateMachine.Event event;
+        /**
+         * The name to be used when persisting this class, or null if this class should not be persisted.
+         */
+        public String getPersistedName() {
+            return null;
+        }
```
Comment (martin-morek) [inline android/src/main/java/io/ably/lib/push/ActivationStateMachine.java:167]: My idea here was to force every event class to provide its own implementation of `getPersistedName()`, that why it was abstract. Main reason for that was avoid bug in the future if we add new type of event and forget to override name.

Comment (QuintinWillison) [reply android/src/main/java/io/ably/lib/push/ActivationStateMachine.java:167]: I understand that and I recognise your point. I also appreciate very much that you are thinking in terms of future maintainability. 😍 

In this particular case, however, I believe that the balance of reduced noise in subclass implementations for what is a well established area of the codebase tips the balance in favour of providing this default implementation.

**Review by QuintinWillison — APPROVED** [review]

**Pull request merged**